In [ ]:
import pandas as pd
import numpy as np

In [ ]:
## Reading csv
data = pd.read_csv('MScFE560_FM_GWP1_Data.csv')
avg_returns = data[:1]
std_dev = data[1:2]
cov_matrix = data[6:]
std_dev

Unnamed: 0         AAPL         AMZN        NFLX         META        GOOG  \
1     St Dev  0.009320453  0.009556763  0.01340882  0.011581856  0.00837577   

   Unnamed: 6                           Unnamed: 7  
1         NaN  Standard deviation of daily returns

In [ ]:
data

Unnamed: 0         AAPL          AMZN          NFLX          META  \
0        Mean  0.044857944   0.010361979  -0.029674829  -0.006934285   
1      St Dev  0.009320453   0.009556763    0.01340882   0.011581856   
2        Skew  -0.33517305  -0.312865226  -3.101653581  -1.936505944   
3        Kurt  5.034092588   4.729806047   41.96801484   23.28923563   
4         NaN          NaN           NaN           NaN           NaN   
5         NaN         AAPL          AMZN          NFLX          META   
6        AAPL            1   0.660738586   0.460041141   0.594784944   
7        AMZN  0.660738586             1   0.593812086   0.626197627   
8        NFLX  0.460041141   0.593812086             1   0.515892088   
9        META  0.594784944   0.626197627   0.515892088             1   
10      GOOGL  0.698431255   0.679322691   0.492697083   0.668023785   

            GOOG  Unnamed: 6                           Unnamed: 7  
0    0.029035134         NaN                 Average daily return  
1     0.00837577         NaN  Standard deviation of daily returns  
2   -0.197699834         NaN            Skewness of daily returns  
3    4.310171108         NaN            Kurtosis of daily returns  
4            NaN         NaN                                  NaN  
5          GOOGL         NaN                                  NaN  
6    0.698431255         NaN  Correlation Matrix of daily returns  
7    0.679322691         NaN                                  NaN  
8    0.492697083         NaN                                  NaN  
9    0.668023785         NaN                                  NaN  
10             1         NaN                                  NaN

In [ ]:
#avg_returns

Unnamed: 0         AAPL         AMZN          NFLX          META  \
0       Mean  0.044857944  0.010361979  -0.029674829  -0.006934285   

          GOOG  Unnamed: 6            Unnamed: 7  
0  0.029035134         NaN  Average daily return

In [ ]:
cov_matrix['rows'] = ['AAPL','AMZN','NFLX','META','GOOG']
cov_matrix

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Unnamed: 0         AAPL         AMZN         NFLX         META  \
6        AAPL            1  0.660738586  0.460041141  0.594784944   
7        AMZN  0.660738586            1  0.593812086  0.626197627   
8        NFLX  0.460041141  0.593812086            1  0.515892088   
9        META  0.594784944  0.626197627  0.515892088            1   
10      GOOGL  0.698431255  0.679322691  0.492697083  0.668023785   

           GOOG  Unnamed: 6                           Unnamed: 7  rows  
6   0.698431255         NaN  Correlation Matrix of daily returns  AAPL  
7   0.679322691         NaN                                  NaN  AMZN  
8   0.492697083         NaN                                  NaN  NFLX  
9   0.668023785         NaN                                  NaN  META  
10            1         NaN                                  NaN  GOOG

## Analysis for Portfolio A

In [ ]:
stocks = ['AAPL','AMZN','NFLX','META','GOOG']

In [ ]:
for i in stocks:
    for j in stocks:
        if(i==j):
            continue
        best_coeff_var = 10**8
        best_wa = 0
        best_wb = 0
        best_std = 0
        best_ret = 0
        for wa in np.arange(-2,1,0.01):
            wb = 1 - wa
            if(max(wa,wb)<=0):
                continue
            expected_return = wa * float(avg_returns[i]) + wb * float(avg_returns[j])
            if(expected_return<=0):
                continue
            prt_std_dev = (wa**2 * float(std_dev[i])**2 + wb**2 * float(std_dev[j])**2 + 2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]))**0.5
            coeff_variation = prt_std_dev / expected_return
            if(coeff_variation < best_coeff_var):
                best_coeff_var = coeff_variation
                best_wa = wa
                best_wb = wb
                best_ret = expected_return
                best_std = prt_std_dev
        print(f"With wa = {best_wa} and wb = {best_wb} weights, with expected returns {best_ret} and with std_dev {best_std}  has the best coeff_var for {i} and {j} and is {best_coeff_var}")
    

With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights, with expected returns 0.0445129843500001 and with std_dev 0.009290670631415986  has the best coeff_var for AAPL and AMZN and is 0.20871821485534628
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights, with expected returns 0.04411261627000022 and with std_dev 0.009289697503383966  has the best coeff_var for AAPL and NFLX and is 0.21059049063253216
With wa = 0.9900000000000029 and wb = 0.009999999999997122 weights, with expected returns 0.04434002171000015 and with std_dev 0.009296601836229055  has the best coeff_var for AAPL and META and is 0.20966615436122715
With wa = 0.9500000000000028 and wb = 0.04999999999999716 weights, with expected returns 0.04406680350000004 and with std_dev 0.009151834430599895  has the best coeff_var for AAPL and GOOG and is 0.2076809231375243
With wa = -0.9999999999999991 and wb = 1.9999999999999991 weights, with expected returns 0.07935390899999996 and with std_dev 0.0142617

#### With wa = -1.1999999999999993 and wb = 2.1999999999999993 weights, with expected returns 0.07219843679999997 and with std_dev 0.013803773147851396  has the best coeff_var for META and GOOG and is 0.19119213323260487


## For portfolio B


In [ ]:
for i in stocks:
    for j in stocks:
        if(i==j):
            continue
        best_coeff_var = 10**8
        best_wa = 0
        best_wb = 0
        best_ret = 0
        best_dev = 0
        for wa in np.arange(0.01,1,0.01):
            wb = 1 - wa
            if(max(wa,wb)<=0):
                continue
            expected_return = wa * float(avg_returns[i]) + wb * float(avg_returns[j])
            if(expected_return<=0):
                continue
            prt_std_dev = (wa**2 * float(std_dev[i])**2 + wb**2 * float(std_dev[j])**2 + 2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]))**0.5
            coeff_variation = prt_std_dev / expected_return
            if(coeff_variation < best_coeff_var):
                best_coeff_var = coeff_variation
                best_wa = wa
                best_wb = wb
                best_dev = prt_std_dev
                best_ret = expected_return
        print(f"With wa = {best_wa} and wb = {best_wb} weights,  with expected returns {best_ret} and with std_dev {best_dev} has the best coeff_var for {i} and {j} and is {best_coeff_var}")
    

With wa = 0.99 and wb = 0.010000000000000009 weights,  with expected returns 0.04451298435 and with std_dev 0.009290670631415978 has the best coeff_var for AAPL and AMZN and is 0.20871821485534653
With wa = 0.99 and wb = 0.010000000000000009 weights,  with expected returns 0.044112616269999995 and with std_dev 0.009289697503383957 has the best coeff_var for AAPL and NFLX and is 0.21059049063253302
With wa = 0.99 and wb = 0.010000000000000009 weights,  with expected returns 0.04434002171 and with std_dev 0.009296601836229048 has the best coeff_var for AAPL and META and is 0.2096661543612277
With wa = 0.9500000000000001 and wb = 0.04999999999999993 weights,  with expected returns 0.0440668035 and with std_dev 0.009151834430599885 has the best coeff_var for AAPL and GOOG and is 0.20768092313752426
With wa = 0.01 and wb = 0.99 weights,  with expected returns 0.04451298435 and with std_dev 0.009290670631415978 has the best coeff_var for AMZN and AAPL and is 0.20871821485534653
With wa = 0.9

## With wa = 0.05 and wb = 0.95 weights,  with expected returns 0.044066803499999994 and with std_dev 0.009151834430599885 has the best coeff_var for GOOG and AAPL and is 0.2076809231375243


# Portfolio C

In [ ]:
for i in stocks:
    for j in stocks:
        for k in stocks:
            if(i==j or j==k or i==k):
                continue
            best_coeff_var = 10**8
            best_wa = 0
            best_wb = 0
            best_wc = 0
            best_dev = 0
            best_ret = 0
            for wa in np.arange(-2,1,0.01):
                for wb in np.arange(-2,1,0.01):    
                    wc = 1 - wa - wb
                    if(max(wa,wb,wc)<=0):
                        continue
                    expected_return = wa * float(avg_returns[i]) + \
                                      wb * float(avg_returns[j]) + \
                                      wc * float(avg_returns[k])

                    if(expected_return<=0):
                        continue
                    prt_std_dev = (wa**2 * float(std_dev[i])**2 + 
                                   wb**2 * float(std_dev[j])**2 +
                                   wc**2 * float(std_dev[k])**2 + 
                                   2*wa*wb*float(std_dev[i])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==i][j]) + 
                                   2*wa*wc*float(std_dev[i])*float(std_dev[k])*float(cov_matrix[cov_matrix.rows==i][k]) +
                                   2*wc*wb*float(std_dev[k])*float(std_dev[j])*float(cov_matrix[cov_matrix.rows==k][j])
                                   )**0.5
                    coeff_variation = prt_std_dev / expected_return
                    if(coeff_variation < best_coeff_var):
                        best_coeff_var = coeff_variation
                        best_wa = wa
                        best_wb = wb
                        best_wc = wc
                        best_dev= prt_std_dev
                        best_ret = expected_return
            print(f"With wa = {best_wa}, wb = {best_wb} and wc = {best_wc} weights,  with expected returns {best_ret} and with std_dev {best_dev} has the best coeff_var for {i},{j} and {k}, and is {best_coeff_var}")
    

With wa = 0.9900000000000029, wb = 0.6600000000000024 and wc = -0.6500000000000052 weights,  with expected returns 0.0705369095500003 and with std_dev 0.011769331570850578 has the best coeff_var for AAPL,AMZN and NFLX, and is 0.16685351890144623
With wa = 0.9900000000000029, wb = 0.44000000000000217 and wc = -0.43000000000000504 weights,  with expected returns 0.05195037787000019 and with std_dev 0.009899898346149577 has the best coeff_var for AAPL,AMZN and META, and is 0.19056451082844725
With wa = 0.9900000000000029, wb = -0.7399999999999989 and wc = 0.749999999999996 weights,  with expected returns 0.05851785060000003 and with std_dev 0.01043998456880858 has the best coeff_var for AAPL,AMZN and GOOG, and is 0.1784068358930561
With wa = 0.9900000000000029, wb = -0.6499999999999988 and wc = 0.6599999999999959 weights,  with expected returns 0.07053690955000005 and with std_dev 0.011769331570850534 has the best coeff_var for AAPL,NFLX and AMZN, and is 0.1668535189014462
With wa = 0.990

## With wa = -1.0399999999999991, wb = -0.9899999999999991 and wc = 3.0299999999999985 weights,  with expected returns 0.1736463346299999 and with std_dev 0.023257681156934735 has the best coeff_var for NFLX,META and AAPL, and is 0.13393706931097316
